src/ZkFold/Symbolic/Data/ByteString.hs

```diff
@@ -1,69 +1,206 @@
+
 
 instance (FromConstant Natural a, Finite a, KnownNat n) => FromConstant Integer (ByteString n a) where
     fromConstant = fromConstant . naturalFromInteger . (`Haskell.mod` (2 ^ getNatural @n))
 
+
+instance (KnownNat p, KnownNat n) => Iso (ByteString n (Zp p)) (UInt n (Zp p)) where
+    from bs@(ByteString r rs)
+      | null rs && ((highRegisterSize @(Zp p) @n) >= (getNatural @n)) = UInt [] r
+      | otherwise = fromConstant @Natural . toConstant $ bs
+
+instance (KnownNat p, KnownNat n) => Iso (UInt n (Zp p)) (ByteString n (Zp p)) where
+    from ui@(UInt rs r)
+      | null rs = ByteString r rs -- A ByteString's high register always has at least the same capacity as UInt's
+      | otherwise = fromConstant @Natural . toConstant $ ui
+
+
 instance (KnownNat p, KnownNat n) => Arbitrary (ByteString n (Zp p)) where
     arbitrary = ByteString
         <$> toss (highRegisterBits @(Zp p) @n)
         <*> replicateA (minNumberOfRegisters @(Zp p) @n -! 1) (toss $ maxBitsPerRegister @(Zp p) @n)
         where toss b = fromConstant <$> chooseInteger (0, 2 ^ b - 1)
 
 
---------------------------------------------------------------------------------
+instance (KnownNat p, KnownNat n) => ShiftBits (ByteString n (Zp p)) where
+    shiftBits b s = fromConstant $ shift (toConstant @_ @Natural b) (Haskell.fromIntegral s) `Haskell.mod` (2 Haskell.^ (getNatural @n))
+
+    -- | @Data.Bits.rotate@ works exactly as @Data.Bits.shift@ for @Natural@, we have from rotate bits manually.
+    rotateBitsL b s
+      | s == 0 = b
+       -- Rotations by k * n + p bits where n is the length of the ByteString are equivalent from rotations by p bits.
+      | s >= (getNatural @n) = rotateBitsL b (s `Haskell.mod` (getNatural @n))
+      | otherwise = fromConstant $ d + m
+        where
+            nat :: Natural
+            nat = toConstant b
+
+            bound :: Natural
+            bound = 2 Haskell.^ getNatural @n
+
+            d, m :: Natural
+            (d, m) = (nat `shiftL` Haskell.fromIntegral s) `Haskell.divMod` bound
+
+    rotateBitsR b s
+      | s == 0 = b
+       -- Rotations by k * n + p bits where n is the length of the ByteString are equivalent from rotations by p bits.
+      | s >= (getNatural @n) = rotateBitsR b (s `Haskell.mod` (getNatural @n))
+      | otherwise = fromConstant $ d + m
+        where
+            nat :: Natural
+            nat = toConstant b
+
+            intS :: Haskell.Int
+            intS = Haskell.fromIntegral s
+
+            m :: Natural
+            m = (nat `Haskell.mod` (2 Haskell.^ intS)) `shiftL` (Haskell.fromIntegral (getNatural @n) Haskell.- intS)
+
+            d :: Natural
+            d = nat `shiftR` intS
+
 
 instance (KnownNat p, KnownNat n) => BoolType (ByteString n (Zp p)) where
     false = fromConstant (0 :: Natural)
 
-    -- | A ByteString with all bits set to 1 is the unity for bitwise and.
+    -- | A ByteString with all bits set from 1 is the unity for bitwise and.
```

Comment:
Typo (after renaming `to -> from`, I guess)

Reply:
oops